xapian-letor/include/xapian-letor/ranker.h

```diff
@@ -23,62 +23,152 @@
+     *  assign scores to the documents based of Learning-to-Rank model. The file is created in the standard format of Letor training file
+     *  as below:
+     *
+     *  0 qid:102 1:0.130742 2:0.000000 3:0.333333 4:0.000000 ... 18:0.750000 19:1.000000 #docid = 13566007
+     *  1 qid:102 1:0.593640 2:1.000000 3:0.000000 4:0.000000 ... 18:0.500000 19:0.000000 #docid = 0740276
+     *
+     *  where first column is relevance judgement of the document with docid as shown in the last column. The second column is query id and in between
+     *  there are 19 feature values.
+     *
+     *  @param  db_path		Path to Xapian::Database to be used.
+     *  @param  query_file      Here you have to give a path to the file (in free text form)  containing training queries in specified format.
+     *  @param  qrel_file       Here supply the path to the qrel file (in free text form) containing the relevance judgements for the
+     *          queries in the training file. This file should be in standard format specified.
+     *  @param  msetsize   This is the mset size used for the first retrieval for training queries. It should be selected depending on the qrel file
+     *          and database size.
+     *  @param  filename   Filename path where the training file has to be stored. Default is "./training-data.txt".
+     *  @param  flist      Xapian::FeatureList object definining what set of features to use for preparing the training file.
+     *          It is initialised by DEFAULT set of Features by default.
+     *          To use a custom set of features, pass a customised Xapian::FeatureList object.
+     *  @exception FileNotFoundError will be thrown if file not found at supplied path
+     *  @exception LetorParseError will be thrown if query file or qrel file could not be parsed
+     */
+    XAPIAN_VISIBILITY_DEFAULT
+    void prepare_training_file(const std::string & db_path,
+			       const std::string & query_file,
+			       const std::string & qrel_file,
+			       Xapian::doccount msetsize,
+			       const std::string & filename = "training-data.txt",
+			       const Xapian::FeatureList & flist = * new Xapian::FeatureList());
+
 class XAPIAN_VISIBILITY_DEFAULT Ranker : public Xapian::Internal::intrusive_base {
-    std::vector<Xapian::FeatureVector> traindata;
+    /// Path to Xapian::Database instance to be used.
+    std::string db_path;
+    /// Xapian::Query to be ranked using Ranking model.
+    Xapian::Query letor_query;
 
   public:
-
     /// Default constructor
     Ranker();
 
     /// Virtual destructor since we have virtual methods.
     virtual ~Ranker();
 
-    /// Returns a vector of RankLists i.e. the training data
-    std::vector<Xapian::FeatureVector> get_traindata() const;
+    /** Specify path to Xapian::Database to be used for Ranker methods.
+     *  @param  dbpath      Path to Xapian::Database to be used.
+     */
+    void set_database_path(const std::string & dbpath);
```

Comment:
Yes, I need a `WritableDatabase` to be able to save model as db metadata, and `Database` for everything else.